### PR TITLE
Add HaveElementWithValue

### LINKED ISSUE #39: HaveElementWithValue?
Hi,

I wish there was a way to assert a `JObject` contains an element with a value as specified.

```c#
// arrange
...
var expectedTokenKey = "token";
var expectedTokenValue = "1234";

// act
JObject json = await ...GetJsonAsync();

// assert
json.Should().HaveElementWithValue(expectedTokenKey, expectedTokenValue);
```

Is there another better way for that I'm unaware of?

### PR DESCRIPTION
**Pull request by weitzhandler**
Fixes #39

### CODE REVIEW

**Comment by weitzhandler** [comment]
Blocker: #41.

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs [hunk 1]
```diff
@@ -662,6 +662,56 @@ public void When_jtoken_does_have_element_NotHaveElement_should_fail()
 
         #endregion (Not)HaveElement
 
+        #region HaveElementAndValue
+
+        [Fact]
+        public void When_jtoken_has_element_and_value_HaveElementAndValue_should_succeed()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:668]: We prefer not to use the exact name of a code element in the test name. I would rephrase this as `When_json_has_the_expected_element_and_value_it_should_succeed`

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs [hunk 2]
```diff
@@ -662,6 +662,56 @@ public void When_jtoken_does_have_element_NotHaveElement_should_fail()
 
         #endregion (Not)HaveElement
 
+        #region HaveElementAndValue
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:665]: What if the subject is `null` Jtoken?

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs [hunk 3]
```diff
@@ -662,6 +662,56 @@ public void When_jtoken_does_have_element_NotHaveElement_should_fail()
...
+        {
+            //-----------------------------------------------------------------------------------------------------------
+            // Arrange
+            //-----------------------------------------------------------------------------------------------------------
+            var subject = JToken.Parse("{ 'id': 42 }");
+
+            //-----------------------------------------------------------------------------------------------------------
+            // Act & Assert
+            //-----------------------------------------------------------------------------------------------------------
+            subject.Should().HaveElementAndValue("id", "42");
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: What if any of the two arguments are empty or `null`?

Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: I guess I'll follow the same rules as `HaveElement` and `HaveValue`. Any better ideas?

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs (hunk 2 above)
Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:665]: Same as [this](https://github.com/fluentassertions/fluentassertions.json/pull/40#discussion_r461743703), no?

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs (hunk 3 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: Don't count on the existing methods to deal with this properly. Let's just make these correct.

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs (hunk 2 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:665]: Don't count on the existing methods to deal with this properly. Let's just make these correct.

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs (hunk 3 above)
Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: I see. What's our agenda then?

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs (hunk 2 above)
Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:665]: Let's resolve this thread and keep our discussion in the other one then.

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs (hunk 3 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: None. I want to make sure new code is properly validating all edge cases.

Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: Right, so what should the expected behavior be on nulls?

Comment (dennisdoomen) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: I guess if the `subject` is `null`, it should throw the usual assertion failure. But if either of the parameters are `null`, I would throw an `ArgumentNullException`. If the first parameter is empty, it should throw an `ArgumentException`. I'm in doubt about the second one, but I guess expecting an empty value is something to be expected. What do you think>

Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: Yep. Makes perfect sense to me.

Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: I think the name parameter makes perfect sense to validate both scenarios, unlike the value parameter which should not be validated, because it makes sense for it to be verified as `null` too.

Comment (weitzhandler) [reply Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:678]: I added some tests to cover these scenarios (https://github.com/fluentassertions/fluentassertions.json/pull/40/commits/50cef9e51c2686628ddd448d7c93b47e2f0a508a):

- `ArgumentNullException` if `expectedElement` is `null`.
- `ArgumentException` if `expectedElement` is empty or whitespace only.
- Verify assertion succeeds if expected and actual JSON values are `null`.


Due to #41 I can't actually run the tests and can only hope for CI to give me the results.

File: Src/FluentAssertions.Json/JTokenAssertions.cs [hunk 4]
```diff
@@ -338,6 +341,37 @@ public AndConstraint<JTokenAssertions> NotMatchRegex(string regularExpression, s
...
+            params object[] becauseArgs)
+        {
+            if (expectedElement == null)
+                throw new ArgumentNullException(nameof(expectedElement));
+            else if (!expectedElement.Trim().Any())
+                throw new ArgumentException("Invalid expected element name.", nameof(expectedElement));
+
+            JToken jToken = Subject[expectedElement];
+            Execute.Assertion
+                .ForCondition(jToken != null && Subject.Value<string>() == expectedValue)
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/JTokenAssertions.cs:367]: What if `expectedValue` is `null`?

File: Src/FluentAssertions.Json/JTokenAssertions.cs [hunk 5]
```diff
@@ -338,6 +341,37 @@ public AndConstraint<JTokenAssertions> NotMatchRegex(string regularExpression, s
...
+        /// </param>
+        public AndWhichConstraint<JTokenAssertions, JToken> HaveElementWithValue(string expectedElement, string expectedValue, string because = "",
+            params object[] becauseArgs)
+        {
+            if (expectedElement == null)
+                throw new ArgumentNullException(nameof(expectedElement));
+            else if (!expectedElement.Trim().Any())
+                throw new ArgumentException("Invalid expected element name.", nameof(expectedElement));
+
+            JToken jToken = Subject[expectedElement];
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/JTokenAssertions.cs:365]: What if `Subject` is `null`?

File: Src/FluentAssertions.Json/JTokenAssertions.cs [hunk 6]
```diff
@@ -338,6 +341,37 @@ public AndConstraint<JTokenAssertions> NotMatchRegex(string regularExpression, s
...
+        /// <param name="becauseArgs">
+        ///     Zero or more objects to format using the placeholders in <see paramref="because" />.
+        /// </param>
+        public AndWhichConstraint<JTokenAssertions, JToken> HaveElementWithValue(string expectedElement, string expectedValue, string because = "",
+            params object[] becauseArgs)
+        {
+            if (expectedElement == null)
+                throw new ArgumentNullException(nameof(expectedElement));
+            else if (!expectedElement.Trim().Any())
+                throw new ArgumentException("Invalid expected element name.", nameof(expectedElement));
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/JTokenAssertions.cs:363]: I think "Please specify a non-null non-empty element name" would be better.

File: Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs [hunk 7]
```diff
@@ -662,6 +662,106 @@ public void When_jtoken_does_have_element_NotHaveElement_should_fail()
...
+            //-----------------------------------------------------------------------------------------------------------
+            // Act & Assert
+            //-----------------------------------------------------------------------------------------------------------
+            subject.Should().Invoking(x => x.HaveElementWithValue("id", "43", "because foo"))
+                .Should().Throw<XunitException>()
+                .WithMessage($@"Expected JSON document {Format(subject)} to have element ""name"" and value ""43"" because foo, but no such element was found.");
+        }
+
+        [Fact]
+        public void When_element_parameter_null_it_should_throw()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Json.Shared.Specs/JTokenAssertionsSpecs.cs:714]: (nitpick) `When_element_argument_is_null_should_throw`

File: Src/FluentAssertions.Json/JTokenAssertions.cs [hunk 8]
```diff
@@ -338,6 +341,37 @@ public AndConstraint<JTokenAssertions> NotMatchRegex(string regularExpression, s
...
+        ///     A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+        ///     is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+        /// </param>
+        /// <param name="becauseArgs">
+        ///     Zero or more objects to format using the placeholders in <see paramref="because" />.
+        /// </param>
+        public AndWhichConstraint<JTokenAssertions, JToken> HaveElementWithValue(string expectedElement, string expectedValue, string because = "",
+            params object[] becauseArgs)
+        {
+            if (expectedElement == null)
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/JTokenAssertions.cs:360]: (nitpick) We prefer braces after `if` statements.

File: Src/FluentAssertions.Json/JTokenAssertions.cs [hunk 9]
```diff
@@ -338,6 +341,37 @@ public AndConstraint<JTokenAssertions> NotMatchRegex(string regularExpression, s
...
+                throw new ArgumentNullException(nameof(expectedElement));
+            else if (!expectedElement.Trim().Any())
```
Comment (jnyrup) [inline Src/FluentAssertions.Json/JTokenAssertions.cs:362]:
```suggestion
            else if (string.IsNullOrWhiteSpace(expectedElement))
```

Comment (weitzhandler) [reply Src/FluentAssertions.Json/JTokenAssertions.cs:362]: We already gathered on line 360 that `expectedElement` is not null.

File: Src/FluentAssertions.Json/JTokenAssertions.cs (hunk 4 above)
Comment (weitzhandler) [reply Src/FluentAssertions.Json/JTokenAssertions.cs:367]: Then it'll verify that the `JSON` value is `null`.

File: Src/FluentAssertions.Json/JTokenAssertions.cs (hunk 9 above)
Comment (jnyrup) [reply Src/FluentAssertions.Json/JTokenAssertions.cs:362]: True, but it's unnecessary to allocate a new string using `Trim` to check if it *contains* whitespace.

**Comment by dennisdoomen** [comment]
We've fixed the build problems and converted the test project into a cross-compiling .NET Core style project. Can you please rebase your changes?

**Comment by weitzhandler** [comment]
Well now that I can actually build and debug this on my machine, I see that it's much easier to use `HaveElementWithValue<T>(string elementKey, T value);`,

Replacing the tests shouldn't be too hard either.
I understand I'll need some additional tests, but I want to get your green light before committing any work.

**Comment by dennisdoomen** [comment]
After this PR is done, we'll release the package.

**Comment by weitzhandler** [comment]
So are you fine with `HaveElementWithValue<T>(string elementKey, T value);` instead of an arbitrary string value?

**Comment by dennisdoomen** [comment]
> So are you fine with `HaveElementWithValue<T>(string elementKey, T value);` instead of an arbitrary string value?

I think it's weird to have the generic `T` for the value since all other APIs (`HaveValue` and `HaveElement`) take a string.

**Comment by weitzhandler** [comment]
Which is a valid point, but in fact the `HaveElement` itself should have been written this way, and in fact I'd be happy to rewrite it, it would just obviously take longer until I find a nice time spot to take of it.

**Comment by dennisdoomen** [comment]
Can you still provide a string value as an argument to the `T` parameter? I guess the underlying `JToken.Value<T>()` will support that, right?

**Comment by weitzhandler** [comment]
I'm sure it will.
Will make an overload.

**Comment by jnyrup** [comment]
What happens if I invoke it with a custom type?
e.g.
```cs
HaveElementWithValue("key", new MyClass());
```
